### PR TITLE
GF Signup: Fix plan type selector on change handling

### PR DESCRIPTION
**Pull request by jeyip**
Related to https://github.com/Automattic/growth-foundations/issues/296 and https://github.com/Automattic/wp-calypso/pull/84895#issuecomment-1850860431

## Proposed Changes

Logic around plan type selector on change handling is currently suboptimal, and it is manifesting as several different bugs
  * The enter key isn't selecting a plan term in the dropdown. We fix the HTML structure and logic of dropdown items to address this issue
  * The plan type selector dropdown immediately redirects the user when clicked in stepper flows ( see p1704870751451319-slack-C0347E545HR )

## Testing Instructions

* Go to `/start?flags=onboarding%2Finterval-dropdown`
* Make sure the dropdown with term interval appears
* Click on the term selector dropdown and use the keyboard to select a new interval. Use the keyboard enter button to select the new interval. The UI should update accordingly

## Pre-merge Checklist

- [ ] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-ajp-p2)?

### CODE REVIEW

File: client/my-sites/plans-grid/components/plan-type-selector/components/interval-type-dropdown.tsx
Comment (chriskmnds) [inline client/my-sites/plans-grid/components/plan-type-selector/components/interval-type-dropdown.tsx]: A couple of issues I think with this (the first two are blockers for me):

1. This looks like something we ought to abstract out of a `plans-grid` component. This event is something that will exist when the interval dropdown is used outside of Calypso (and I'm guessing the action there would not be to send user to `/start/plans[something]`). 
2. Not sure if this is correct within Calypso either. Do we also send users to `/start/plans` when used from within admin (`/plans` page)?
3. This line reads ambiguously to me: `page( '/start/plans${ selectedItem.url }' );`. You cannot "attach a URL to a URL path", so something's weird/off in what/how `selectedItem.url` has been implemented and defined. Something's weirder when they are all empty strings `''` right now.

**Review by chriskmnds — DISMISSED** [review]
Thanks! Requesting changes from the comment below :-)

File: client/my-sites/plans-grid/components/plan-type-selector/components/interval-type-dropdown.tsx
Comment (jeyip) [reply client/my-sites/plans-grid/components/plan-type-selector/components/interval-type-dropdown.tsx]: I was at the end of the day yesterday and wanted to post something up for discussion before signing off, so I wasn't focused on the url, but thanks for chiming in Christos. Everything you bring up are great points to address.

I'll work on them shortly 🙂

Comment (jeyip) [reply client/my-sites/plans-grid/components/plan-type-selector/components/interval-type-dropdown.tsx]: The most recent set of commits reflect our discussion so far. I added an `onPlanTypeSelectorChange` method to the PlanTypeSelector interface, made generation of the url dynamic to accommodate the `/plans` page, and looked into the `url` being appended to a URL path

I'll be explaining my reasoning soon ( and might make another change or two )

Comment (jeyip) [reply client/my-sites/plans-grid/components/plan-type-selector/components/interval-type-dropdown.tsx]: > This looks like something we ought to abstract out of a plans-grid component. This event is something that will exist when the interval dropdown is used outside of Calypso (and I'm guessing the action there would not be to send user to /start/plans[something]).

I exposed onChange functionality through a callback we add as props to the `PlanTypeSelector` interface. This seems like a reasonable way to allow consumers to define how they'd handle updates of plan type terms, but open to your thoughts

> Not sure if this is correct within Calypso either. Do we also send users to /start/plans when used from within 
admin (/plans page)?

Definitely not 👍 I've amended this for dynamic url generation

> This line reads ambiguously to me: page( '/start/plans${ selectedItem.url }' );. You cannot "attach a URL to a URL path", so something's weird/off in what/how selectedItem.url has been implemented and defined. Something's weirder when they are all empty strings '' right now.

After looking into this, there is, at least to me, unexpected data returned by the the `useIntervalOptions` hook.  The URL attribute might be either a string of coalesced query params in onboarding Ex. `?flags=onboarding%2Finterval-dropdown&intervalType=2yearly` or a relative url in Calypso admin `/plans` Ex. `/plans/2yearly/jeremyyip124.wordpress.com`

I suppose this is a reflection of the fact that plan type terms are updated differently in onboarding and Calypso admin `/plans`. Not ideal, but not entirely sure if there's a graceful way of handling this without taking a deeper look into actual architectural design of plan types. 🤔

File: client/my-sites/plans-grid/components/plan-type-selector/components/interval-type-dropdown.tsx
Comment (jeyip) [inline client/my-sites/plans-grid/components/plan-type-selector/components/interval-type-dropdown.tsx]: Note: Refine this inline comment, or consider making a note of this in an issue

File: client/my-sites/plans-grid/components/plan-type-selector/components/interval-type-dropdown.tsx
Comment (chriskmnds) [reply client/my-sites/plans-grid/components/plan-type-selector/components/interval-type-dropdown.tsx]: > After looking into this, there is, at least to me, unexpected data returned by the the useIntervalOptions hook. The URL attribute might be either a string of coalesced query params in onboarding Ex. ?flags=onboarding%2Finterval-dropdown&intervalType=2yearly or a relative url in Calypso admin /plans Ex. /plans/2yearly/jeremyyip124.wordpress.com

The problem here stems from the very fact that the interval toggle has knowledge of the expected URL structure at the consuming end. This is problematic, more so when that knowledge is only so partial (the consumer defines the route path, then the toggle provides the remaining part for it, whether a query or an actual path).

I'll revisit in a bit after reviewing the changes and see if any action before approving. My immediate impression is we should stop using `option.url`, remove that structure completely from interval toggle.

Also, we really need to make a more conscious effort on how we name things. A partial path or a partial query string is NOT a URL. This seems to be a bit of a recurring issue, this mismatch in semantics and naming.

File: client/my-sites/plans-features-main/index.tsx
Comment (jeyip) [inline client/my-sites/plans-features-main/index.tsx]: `Note`:

If we go with this approach, it's probably worthwhile to add an inline comment describing that `url` can be a query string instead of a url

File: client/my-sites/plans-features-main/index.tsx
Comment (jeyip) [inline client/my-sites/plans-features-main/index.tsx]: `Note`:

DRY `selectedItem.url` before merging

File: client/my-sites/plans-grid/components/plan-type-selector/components/interval-type-dropdown.tsx
Comment (ddc22) [reply client/my-sites/plans-grid/components/plan-type-selector/components/interval-type-dropdown.tsx]: > The problem here stems from the very fact that the interval toggle has knowledge of the expected URL

This is true, there is a lot of unrelated params included in this area of the code that needs to be cleaned up.
https://github.com/Automattic/wp-calypso/blob/b63dede300b78dda7df28eba5c12524c5a1e8e03/client/my-sites/plans-grid/components/plan-type-selector/utils.ts#L35-L38

The hook above just generates these urls using `generatePath`.

> My immediate impression is we should stop using option.url

The main blocker for moving this out earlier was the behaviour of segment control component which used anchor tags and its ability to refresh query params, trigger a url change and a prop change without a page refresh. (Or so I thought). With @jeyip's discovery here it seems like we can finally do this cleanup 🙌 .

The plan type selector can just take `displayedIntervalTypes` and an `onIntervalChange` callback. Some of that logic is introduced here https://github.com/Automattic/wp-calypso/pull/86192.

> A partial path or a partial query string is NOT a URL

This is true and this naming is my fault. I incorrectly assumed early on that what's being generated is a URL and failed to correct it later. Sorry for the confusion caused.

Comment (jeyip) [reply client/my-sites/plans-grid/components/plan-type-selector/components/interval-type-dropdown.tsx]: It's a bit past the end of my day. I somewhat understand `The plan type selector can just take displayedIntervalTypes and an onIntervalChange callback.` @jdc91, so if you have time to elaborate on this, that'd be great. If not, no worries -- I'm sure I can glean more of the meaning tomorrow when I return to this 🙂

Comment (jeyip) [reply client/my-sites/plans-grid/components/plan-type-selector/components/interval-type-dropdown.tsx]: @chriskmnds + @jdc91 what I understand from this conversation is that we want to do a few things here:

- Simplify the plan type selector and make it more of a purely presentational component by shifting any url generation concerns to consuming components
- Practically speaking, that means refactoring the `interval-type-dropdown`, and, eventually, the `interval-type-toggle` as well.
  - Move any relative path or query string generation from either plan type selector to Plans Features Main
  - Expose a callback that's invoked on plan type selection. Pass simple keys to callback and allow consuming components to decide what to do with the information
  - Bonus points? Align with the refactoring effort by scooting plan type selector UI logic into wpcom-plans-ui data store

I'll start moving in this direction, but if any part of this feels off base, please feel free to chime in 🙂

Comment (chriskmnds) [reply client/my-sites/plans-grid/components/plan-type-selector/components/interval-type-dropdown.tsx]: That sounds one hundred percent correct. Nicely outlined. Thank you both :-)

File: client/my-sites/plans-features-main/index.tsx
Comment (jeyip) [reply client/my-sites/plans-features-main/index.tsx]: No longer applicable as we refactor https://github.com/Automattic/wp-calypso/pull/86127#discussion_r1449253021

File: client/my-sites/plans-grid/components/plan-type-selector/components/interval-type-dropdown.tsx
Comment (jeyip) [reply client/my-sites/plans-grid/components/plan-type-selector/components/interval-type-dropdown.tsx]: > Practically speaking, that means refactoring the interval-type-dropdown, and, eventually, the interval-type-toggle as well.
> - Move any relative path or query string generation from either plan type selector to Plans Features Main
> - Expose a callback that's invoked on plan type selection. Pass simple keys to callback and allow consuming components to decide what to do with the information
> - Bonus points? Align with the refactoring effort by scooting plan type selector UI logic into wpcom-plans-ui data store

Quick update:

I've focused this PR on updating _only_ the plan type **dropdown**. ( I will handle the plan type **toggle** in an immediate follow-up ) In doing so, I've made several important changes:
- Path / Query string generation for the plan type selector lives [here](https://github.com/Automattic/wp-calypso/blob/trunk/client/my-sites/plans-grid/components/plan-type-selector/utils.ts). In the spirit of lifting logic outside of the plan type selector component, I duplicate this code into [a new hook usePlanTypeDestinationCallback]( https://github.com/Automattic/wp-calypso/pull/86127/files#diff-1cd50998bfbda4ec843e2af91f7b429e9fcdf0c2faf42c5e3a3715f2f77d538cR15 ) in Plans Features Main
- The plan type **toggle** still relies on the [plan type selector utils file](https://github.com/Automattic/wp-calypso/blob/trunk/client/my-sites/plans-grid/components/plan-type-selector/utils.ts) containing Path / Query string generation logic. I will delete the file once I finish refactoring said toggle
- [I delete the options.url structure from useIntervalOptions]( https://github.com/Automattic/wp-calypso/pull/86127/files#diff-70a87a1e3f781a90ba99174199d5162076cd36ef736a9970842a119925f48436L102-L108 ) because it will now be handled outside of the plan type selector
- I expose the [onPlanTypeSelectorChange callback]( https://github.com/Automattic/wp-calypso/pull/86127/files#diff-18214fd84316736ff6662f03fa201b6255f4c20f277c0eb683b5e36e9c9dea6dR867  ) to consumers to allow them to decide how they'd like to handle plan type selection. I've moved path vs. query param processing out into Plans Features Main accordingly

I've left some notes to myself throughout the PR about smaller things to clean up, but the general structure is there and could use some preliminary review from folks @jdc91 + @chriskmnds 🙂

File: client/my-sites/plans-grid/components/plan-type-selector/components/interval-type-dropdown.tsx
Comment (jeyip) [reply client/my-sites/plans-grid/components/plan-type-selector/components/interval-type-dropdown.tsx]: No longer necessary after addressing code review

File: client/my-sites/plans-features-main/index.tsx
Comment (jeyip) [inline client/my-sites/plans-features-main/index.tsx:734]: **Note:**

@jdc91 Looking for a confidence check. I copied L733-749 from [the useIntervalOptionHook]( https://github.com/Automattic/wp-calypso/pull/86127/files#diff-70a87a1e3f781a90ba99174199d5162076cd36ef736a9970842a119925f48436L82-L95 ) which you wrote. There is an `addtionalPathProps` value that reads the `redirectTo` prop from `planTypeSelectorProps`.

From what I can tell, the value never exists on `planTypeSelectorProps`. Could you elaborate on what's happening?

File: client/landing/stepper/declarative-flow/internals/steps-repository/plans/plans-wrapper.tsx
Comment (jeyip) [inline client/landing/stepper/declarative-flow/internals/steps-repository/plans/plans-wrapper.tsx:105]: **Note:**

Stepper uses `react-router` for client side routing as opposed to the `page` library, so it needs its own specific logic to handle plan type selection

File: client/my-sites/plans-grid/components/plan-type-selector/types.ts
Comment (chriskmnds) [inline client/my-sites/plans-grid/components/plan-type-selector/types.ts]: precisely 🙌 
now we have a proper interface that's reusable for any context, without the complexity of "the component knowing too much"

NIT: wonder if a better name might work for this, something more granular or general to reflect the actual operation? `onSomethingChange` - but nothing comes to mind though, so.. 🤷

File: client/my-sites/plans-features-main/hooks/use-plan-type-destination-callback.ts
Comment (chriskmnds) [inline client/my-sites/plans-features-main/hooks/use-plan-type-destination-callback.ts:22]: I wonder (I haven't checked, sorry) if these props are only needed for the callback here and nowhere else in the plan-type-selector component, in which case we can remove them from `PlanTypeSelectorProps` (that we define on the actual component) and minimize that interface a bit.

Sorry if I'm missing something though :-)

File: client/my-sites/plans-features-main/index.tsx
Comment (chriskmnds) [inline client/my-sites/plans-features-main/index.tsx]: Similar to the other comment https://github.com/Automattic/wp-calypso/pull/86127/files#r1450171723
Is `redirectTo` used anywhere beyond `plans-features-main` by the `PlanTypeSelector` components?

File: client/my-sites/plans-grid/components/plan-type-selector/hooks/use-interval-options.tsx
Comment (chriskmnds) [inline client/my-sites/plans-grid/components/plan-type-selector/hooks/use-interval-options.tsx:64]: 🙇

File: client/my-sites/plans-features-main/index.tsx
Comment (ddc22) [reply client/my-sites/plans-features-main/index.tsx:734]: Sadly this was copied directly from the previous toggle component. Maybe it is a pending cleanup but your guess is as good as mine.

https://github.com/Automattic/wp-calypso/blob/f8299cc81ab0653ae3167b33d52683065086f37a/client/my-sites/plans-grid/components/plan-type-selector/components/interval-type-toggle.tsx#L65-L79

**Comment by jeyip** [comment]
Planning to merge this PR once I address all comments and conduct some small clean up tasks I left in inline `TODO` comments

File: client/my-sites/plans-features-main/hooks/use-plan-type-destination-callback.ts
Comment (jeyip) [reply client/my-sites/plans-features-main/hooks/use-plan-type-destination-callback.ts:22]: Great question 👍 I also noticed some oddness with plan type selector props yesterday https://github.com/Automattic/wp-calypso/pull/86127#discussion_r1449739205 ( although it has to do with entirely unused props ). I'll do some quick auditing today before merging

File: client/my-sites/plans-features-main/index.tsx
Comment (jeyip) [reply client/my-sites/plans-features-main/index.tsx]: https://github.com/Automattic/wp-calypso/pull/86127#discussion_r1450955200

File: client/my-sites/plans-grid/components/plan-type-selector/types.ts
Comment (jeyip) [reply client/my-sites/plans-grid/components/plan-type-selector/types.ts]: > wonder if a better name might work for this, something more granular or general to reflect the actual operation?

This makes sense to me since the plan type selector will be part of the plans grid shared package. Perhaps `onPlanTermChange` or `onPlanIntervalChange`... 🤔

File: client/my-sites/plans-features-main/index.tsx
Comment (jeyip) [reply client/my-sites/plans-features-main/index.tsx:734]: Thanks for the heads up -- I'll do some auditing

File: client/my-sites/plans-features-main/hooks/use-plan-type-destination-callback.ts
Comment (jeyip) [reply client/my-sites/plans-features-main/hooks/use-plan-type-destination-callback.ts:22]: `withDiscount`, `selectedFeature`, and `selectedPlan` appear to only be used by the callback. I'll remove them from `PlanTypeSelectorProps` when I refactor the plan type selector toggle in the immediate follow-up I was planning

File: client/my-sites/plans-features-main/index.tsx
Comment (jeyip) [reply client/my-sites/plans-features-main/index.tsx]: It looks like the `redirectTo` props doesn't even exist in PlansFeaturesMain or 😆 Removing references to it entirely in our new hook usePlanTypeDestinationCallback. I'll clean other references up in the plan type selector component in the follow-up where I refactor the Plan Type Selector Toggle

File: client/my-sites/plans-features-main/index.tsx
Comment (jeyip) [reply client/my-sites/plans-features-main/index.tsx:734]: Looks to be unused. Removing references to it

File: client/my-sites/plans-grid/components/plan-type-selector/types.ts
Comment (jeyip) [reply client/my-sites/plans-grid/components/plan-type-selector/types.ts]: I decided on onPlanIntervalChange. If we don't like it, I'm happy to switch back or to something else entirely -- it's an easy adjustment 🙂

**Pull request merged**
